statics/js/topologymanager_package/src/topologymanager/topology_layout/infra/index.ts

```diff
@@ -0,0 +1,117 @@
+import { DataSourceI, DataSourceRegistry } from "../../data_source/index";
+import { TopologyLayoutI } from "../index";
+import LayoutConfig from '../config';
+import * as events from 'events';
+import { DataManager } from '../base/index';
+import { LayoutBridgeUI, LayoutBridgeUII } from '../base/ui/index';
+import { LayoutUI, EdgeUI, NodeUI, GroupUI } from '../base/ui/index';
+import { LabelRetrieveInformationStrategy } from '../base/edge/label/index';
+export default class SkydiveInfraLayout implements TopologyLayoutI {
+    uiBridge: LayoutBridgeUII;
+    dataManager: DataManager = new DataManager();
+    e: events.EventEmitter = new events.EventEmitter();
+    alias: string = "skydive_infra";
+    active: boolean = false;
+    config: LayoutConfig;
+    dataSources: DataSourceRegistry = new DataSourceRegistry();
+    selector: string;
+    constructor(selector: string) {
+        this.selector = selector;
+        this.uiBridge = new LayoutBridgeUI(selector);
+        this.uiBridge.useEventEmitter(this.e);
+        this.uiBridge.useConfig(this.config);
+        this.uiBridge.useDataManager(this.dataManager);
+        this.uiBridge.useLayoutUI(new LayoutUI(selector));
+        this.uiBridge.useNodeUI(new NodeUI());
+        this.uiBridge.useEdgeUI(new EdgeUI());
+        this.uiBridge.useGroupUI(new GroupUI());
+        this.dataManager.useLayoutContext(this.uiBridge.layoutContext);
+    }
+    initializer() {
+        console.log("Try to initialize topology " + this.alias);
+        $(this.selector).empty();
+        this.active = true;
+        this.uiBridge.start();
+    }
+    useLinkLabelStrategy(linkLabelType: string) {
+        const strategy = LabelRetrieveInformationStrategy(linkLabelType);
+        strategy.setup(this.config);
+        this.uiBridge.useLinkLabelStrategy(strategy);
+    }
+    useConfig(config: LayoutConfig) {
+        this.config = config;
+        this.uiBridge.useConfig(this.config);
+    }
+    remove() {
+        this.dataSources.sources.forEach((source: DataSourceI) => {
+            source.unsubscribe();
+        });
+        this.removeLayout();
+    }
+    removeLayout() {
+        this.active = false;
+        this.uiBridge.remove();
+        $(this.selector).empty();
+        this.dataManager.removeOldData();
+        this.e.emit('ui.update');
+    }
+
+    addDataSource(dataSource: DataSourceI, defaultSource?: boolean) {
+        this.dataSources.addSource(dataSource, !!defaultSource);
+    }
+    reactToDataSourceEvent(dataSource: DataSourceI, eventName: string, ...args: Array<any>) {
```

Comment:
I'm not sure about the datasource pattern. Finally layout and datasource are really tied and it seems that the datasource is more level of indirection for broadcasting websocket messages. Layout could just subscribe to a graphdatasource not using an interface that basically does nothing. There is no difference between all data sources except the gremlin filter.

Reply:
> I'm not sure about the datasource pattern. Finally layout and datasource are really tied and it seems that the datasource is more level of indirection for broadcasting websocket messages. 

Yes, but it's mostly about proper layering of the systems. It should be very easy to replace any communication gateway we use in our app. For example, right now I am not happy about websocket WsHandler I copied mostly from previous implementation. But I've no found better alternative. So, I think it is very good layer above websocket communication way to abstract away all websocket internal details of communication.

> Layout could just subscribe to a graphdatasource not using an interface that basically does nothing. There is no difference between all data sources except the gremlin filter.

You are right but it's very important to have an ability to use multiple data sources for layout. For instance, right now I copied link label implementation from previous javascript BUT there we use api to get an information about link latencies, etc
It's bad idea because presentation should know nothing about data sources it gets information from. So, later I'll create a separated data source based on api call to get a link latencies information to get rid of such improper system layering.